### ISSUE TITLE
How to show username in the item cell ,It only shows avatar now

### ISSUE THREAD

**Issue by MrShiY** [badoo/Chatto#479]
Please tell me,Thank you very much

**Comment by neilabdev** [comment]
+1

**Comment by Daxito** [comment]
I accomplished this by adding a new DecoratedChatItem for every message (see ChatItemsDecoratorProtocol)

**Comment by MrShiY** [comment]
What you said was to register ChatItemPresenter Cell,  like timeSeparator and messageSendingStatus ok?

**Comment by neilabdev** [comment]
Hey MrShiY, as I know it takes a while to even get started, I will do my best to give you an easy example to get you in the right direction.

Basically, in my case I subclassed the *BaseChatViewController* in *ChatAdditions* where the decorator handler via property *decoratorItems* is specified.


```swift
class MySubclassChatViewController: BaseChatViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // The line below specifies your custom decorator
        self.chatItemsDecorator = MyChatItemsDecorator(messagesSelector: self.messagesSelector) 
    }
}

```
It should be noted that I literally copied my decorator from the Chatto Demo application, as I wasn't sure what needed to be there. Below its called *MyChatItemsDecorator*.

```swift
class MyChatItemsDecorator: ChatItemsDecoratorProtocol {
    func decorateItems(_ chatItems: [ChatItemProtocol]) -> [DecoratedChatItem] {
func decorateItems(_ chatItems: [ChatItemProtocol]) -> [DecoratedChatItem] {
        var decoratedChatItems = [DecoratedChatItem]()
        let calendar = Calendar.current

        for (index, chatItem) in chatItems.enumerated() {
            let next: ChatItemProtocol? = (index + 1 < chatItems.count) ? chatItems[index + 1] : nil
            let prev: ChatItemProtocol? = (index > 0) ? chatItems[index - 1] : nil

            var bottomMargin = self.separationAfterItem(chatItem, next: next)
            var showsTail = false
            var additionalItems = [DecoratedChatItem]()
            var addTimeSeparator = false
            var isSelected = false
            var isShowingSelectionIndicator = false

            if let currentMessage = chatItem as? MessageModelProtocol {
                if let nextMessage = next as? MessageModelProtocol {
                    showsTail = currentMessage.senderId != nextMessage.senderId
                } else {
                    showsTail = true
                }

                if let previousMessage = prev as? MessageModelProtocol {
                    addTimeSeparator = !calendar.isDate(currentMessage.date, inSameDayAs: previousMessage.date)
                } else {
                    addTimeSeparator = true
                }

                if let myMessage = chatItem as? MyTextMessageModel,
                   let username = paradeMessage.username as? String, showsTail == true {
                    bottomMargin = 0.0. //HERE:  I ensure there is no space after the Avatar/message if its the last message , aka showTail
                    //NOTE: Add username here , determines if the username is on left or right with the .isOwner property in my custom model.
                    additionalItems.append(
                            DecoratedChatItem(
                                    chatItem: UsernameModel(uid: "\(currentMessage.uid)-decoration-username",
                                            username: username, owner: myMessage.isOwner),
                                    decorationAttributes: nil))
                }

                if self.showsStatusForMessage(currentMessage) {
                    additionalItems.append(
                            DecoratedChatItem(
                                    chatItem: SendingStatusModel(uid: "\(currentMessage.uid)-decoration-status", status: currentMessage.status),
                                    decorationAttributes: nil)
                    )
                }

                if addTimeSeparator {
                    let dateTimeStamp = DecoratedChatItem(chatItem: TimeSeparatorModel(uid: "\(currentMessage.uid)-time-separator", date: currentMessage.date.toWeekDayAndDateString()), decorationAttributes: nil)
                    decoratedChatItems.append(dateTimeStamp)
                }

                isSelected = self.messagesSelector.isMessageSelected(currentMessage)
                isShowingSelectionIndicator = self.messagesSelector.isActive && self.messagesSelector.canSelectMessage(currentMessage)
            }

            decoratedChatItems.append(
                    DecoratedChatItem(
                            chatItem: chatItem,
                            decorationAttributes: ChatItemDecorationAttributes(bottomMargin: bottomMargin,
                                    canShowTail: showsTail, canShowAvatar: showsTail, canShowFailedIcon: true)
                    )
            )
            decoratedChatItems.append(contentsOf: additionalItems)
        }

        return decoratedChatItems
    }
    }
}
```

Basically, what is happening is each message is iterated, and the decorator reviews each row providing the ability to add additional items as if they were added to the dataSource specifically. In this case, by virtue of addition of a *Status message*, or *userName*  to the decoratedChatItems that is returned, they are respectively turned into new CollectionViewCell's you specific below each row, etc.

While its a big pill to swallow, copying the implementation of each decorated item in the *Chatto* app could be of help as I hope this minor example might be. Additional applicable links are below: Thanks!


Decorator explanation:
https://github.com/badoo/Chatto/wiki/Decorator


How to create a new cell (in your case, username)
https://github.com/badoo/Chatto/wiki/Tutorial_Add_Custom_Cell

**Comment by MrShiY** [comment]
Very cool, thank neilabdev and Daxito for your help
